ledger-ondisk/src/LedgerOnDisk/Diff.hs

```diff
@@ -38,29 +38,72 @@ data Diff v where
 deriving stock instance Eq v => Eq (Diff v)
 deriving stock instance Show v => Show (Diff v)
 
+{-  (<>) on Diff is not commutative, so I think we should change the order of arguments here for 2 reasons
+ 1) I like to think of (d1 <> d2) as composition of changes. So in accordance with Haskell's composition
+    we should think of addpying the d2 first.  E.g. in Haskell (f . g) x == f (g x). So apply (x <> y) x == apply x (apply y x)
+ 2) When we lift (Diff v) to Map k (Diff v), we will want to use Map.insertWith (<>) k v x, and Map.unionWith (<>) m1 m2
+    and these require to us apply the change on the right first.
+So it is best to change the order.
+
 instance Semigroup (Diff v) where
   x <> DNoChange = x
   DMappend x <> DMappend y = DMappend $ x <> y
   DChangeTo x <> DMappend y = DChangeTo $ x <> y
   DRemove <> DMappend y = DChangeTo y
   _ <> y = y
 
+Another thing, is that DRemove (applied first) cancels a subsequent (DMappend v) so the 4th line
+when appying things in the eft first should be
+   DRemove <> DMappend y = DRemove.
+We can also make the function clearer by making it more verbose, so it is easy to see all the cases are covered.
+So I suggest this.
```

Comment:
I think either of
```
DRemove <> DMappend _ = DRemove
```
or
```
DRremove <> DMappend x = DChangeTo x
```
are valid instances. Am I wrong? Otherwise can you elaborate on why you prefer the first?

Reply:
You don't really see the difference till you lift the Diff to Maps.
I want  (apply (d1 <> d2) m)  to be  apply d1 (apply d2 m)
so at the map level suppose d1 was { 'a' -> Dremove} and d2 was {'a' -> DMappend 4}  and m = {'a' -> 3, 'b' -> 5}
So     apply d1 (apply d2 m)  goes in two steps {'a' -> 3, 'b' -> 5}  apply d2 {'a' -> 7, 'b' -> 5} apply d2  {'b' -> 5}

Now  under the second interpretation of <> for Diff, we have at the map level  {'a' -> Dremove} <> {'a' -> Mppend 4} goes to
{'a' -> DChangeTo '4'}   and now       (apply (d1 <> d2) m)  goes to   apply {'a' -> DChangeTo '4'} {'a' -> 3, 'b' -> 5} goes to
{'a' -> 7, 'b' -> 5},  under the first interpretation we get {'b' -> 5}

So the law     (apply (d1 <> d2) m)  == apply d1 (apply d2 m)  No longer holds. In the end, we deveop all his to work at the map level, so the first interpretation is what we want.

Reply:
I agree that it is very important that `(apply (d1 <> d2) m) == apply d1 (apply d2 m)` holds, up to I'm used to thinking of `d1 <> d2` applying d1 then d2. In this note I'm going to use `<>` for the semigroup operator that applies `d2` first and `++` for the semigroup operator that applies `d1` first to try to remove ambiguity.

I think there are two adjustments we're talking about:
* The change so that `<>` is the semigroup operator rather than the current `++`
* The change so that `DMappend x <> DRemove == DRemove` rather than the current `DMappend x <> DRemove == DChangeTo x`

Let's asssume that we have applied change 1 but not change 2. Now, setting 
```haskell
d1 = {'a' -> DRemove}
d2 = {'a' -> DMappend 4}
m = {'a' -> 3, 'b' -> 5}
```
we have
```
d1 <> d2 == {'a' -> DRemove }
apply d2 m == {'a' -> 7, 'b' -> 5 }
apply d1 (apply d2 m) == {'b' -> 5 }
apply (d1 <> d2) == {'b' -> 5}
```
So this is not a counteraxample.

If we try
````haskell
d2 <> d1 == { 'a' -> DChangeTo 4}
apply d1 m == {'b' -> 5 }
apply d2 (apply d1 m) == {'a' -> 4, 'b' -> 5}
apply (d2 <> d1) m == {'a' -> 4, 'b' -> 5}
```
So the law holds here too. I think either interpretation of `DMappend x <> DRemove` satisfies our law.